docs/painless/painless-basic-syntax.asciidoc

```diff
@@ -0,0 +1,304 @@
+`float -90.0`, `int -18` in octal, and `int 3882` in hex.
+
+[source,Java]
+----
+0
+0D
+1234L
+-90f
+-022
+0xF2A
+----
+
+[[floats]]
+===== Floats
+
+Use floating point literals to specify a floating point value of the
+<<primitive-types, primitive types>> float or double. Use the following single
+letter designations for the <<primitive-types, primitive type>>: `f` for
+`float` and `d` for `double`. If not specified, the type defaults to `double`.
+
+*Grammar:*
+[source,ANTLR4]
+----
+DECIMAL: '-'? ( '0' | [1-9] [0-9]* ) (DOT [0-9]+)? EXPONENT? [fFdD]?;
+EXPONENT: ( [eE] [+\-]? [0-9]+ );
+----
+
+*Examples:*
+
+Floating point literals of `double 0.0`, `double 1000000.0` in
+exponent notation, `double 0.977777`, `double -126.34`, and `float 89.9`.
+
+[source,Java]
+----
+0.0
+1E6
+0.977777
+-126.34
+89.9F
+----
+
+[[strings]]
+===== Strings
+
+Use string literals to specify string values of the
+<<string-type, String type>> with either single or double quotes.
+Use a `\"` token to include a double-quote as part of a double-quoted string
+literal. Use a `\'` token to include a single-quote as part of a single-quoted
+string literal.  Use a `\\` token to include a backslash as part of any string
+literal.
+
+*Grammar:*
+[source,ANTLR4]
+----
+STRING: ( '"'  ( '\\"'  | '\\\\' | ~[\\"] )*? '"'  )
+      | ( '\'' ( '\\\'' | '\\\\' | ~[\\'] )*? '\'' );
+----
+
+*Examples:*
+
+String literals using both single-quotes and double-quotes.
+
+[source,Java]
+----
+"double-quoted String literal"
+'single-quoted String literal'
+"\"double-quoted String with escaped double-quotes\" and backslash: \\"
+'\'single-quoted String with escaped single-quotes\' and backslash \\'
+"double-quoted String with non-escaped 'single-quotes'"
+'single-quoted String with non-escaped "double-quotes"'
+----
+
+[[characters]]
+===== Characters
+
+Use the <<casting, casting operator>> to convert string literals or
+<<string-type, String type>> values into <<primitive-types, char type>> values.
+<<string-type, String type>> values converted into
+<<primitive-types, char type>> values must be exactly one character in length
+or an error will occur.
+
+*Examples:*
+
+Casting string literals into <<primitive-types, char type>> values.
+
+[source,Java]
+----
+(char)"C"
+(char)'c'
+----
+
+Casting a <<string-type, String type>> value into a
+<<primitive-types, char type>> value.
+
+[source,Java]
+----
+String s = "s";
+char c = (char)s;
+----
+
+[[variables]]
+==== Variables
+
+Variables in Painless must be declared and can be statically or <<dynamic-types,
+dynamically typed>>.
+
+[[variable-identifiers]]
+===== Variable Identifiers
+
+Specify variable identifiers using the following grammar. Variable identifiers
+must start with a letter or underscore. You cannot use <<keywords, keywords>> or
+<<painless-types, types>> as identifiers.
+
+*Grammar:*
+[source,ANTLR4]
+----
+ID: [_a-zA-Z] [_a-zA-Z-0-9]*;
+----
+
+*Examples:*
+[source,Java]
+----
+_
```

Comment:
I wonder if we should document that `_` is a valid variable identifier. I think Java has been trying to move away from that.

Reply:
Removed from the docs.  Probably not worth removing from the language at this point given backwords compat issues, but that's a different discussion.